core/src/main/java/org/apache/calcite/rel/rules/FilterWindowTransposeRule.java

```diff
@@ -0,0 +1,131 @@
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelOptUtil;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Filter;
+import org.apache.calcite.rel.core.Window;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.util.ImmutableBitSet;
+
+import com.google.common.collect.ImmutableList;
+
+import org.immutables.value.Value;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Planner rule that pushes a {@link org.apache.calcite.rel.core.Filter}
+ * past a {@link org.apache.calcite.rel.core.Window}.
+ *
+ * <p> If {@code Filter} condition used columns belongs {@code Window} partition keys,
+ * then we could push the condition past the {@code Window}.
+ *
+ * <p> For example:
+ *  <blockquote><pre>{@code
+ *    LogicalProject(NAME=[$0], DEPTNO=[$1], EXPR$2=[$2])
+ *     LogicalFilter(condition=[>($1, 0)])
+ *       LogicalProject(NAME=[$1], DEPTNO=[$0], EXPR$2=[$2])
+ *         LogicalWindow(window#0=[window(partition {0} aggs [COUNT()])])
+ *           LogicalTableScan(table=[[CATALOG, SALES, DEPT]])
+ *  }</pre></blockquote>
+ *
+ * <p> will convert to:
+ * <blockquote><pre>{@code
+ *    LogicalProject(NAME=[$1], DEPTNO=[$0], EXPR$2=[$2])
+ *      LogicalWindow(window#0=[window(partition {0} aggs [COUNT()])])
+ *        LogicalFilter(condition=[>($0, 0)])
+ *          LogicalTableScan(table=[[CATALOG, SALES, DEPT]])
+ * }</pre></blockquote>
+ *
+ * @see CoreRules#FILTER_PROJECT_TRANSPOSE
+ * @see CoreRules#PROJECT_TO_LOGICAL_PROJECT_AND_WINDOW
+ * @see CoreRules#FILTER_WINDOW_TRANSPOSE
+ */
+@Value.Enclosing
+public class FilterWindowTransposeRule
+    extends RelRule<FilterWindowTransposeRule.Config>
+    implements TransformationRule {
+
+  protected FilterWindowTransposeRule(final Config config) {
+    super(config);
+  }
+
+  @Override public void onMatch(final RelOptRuleCall call) {
+    final Filter filterRel = call.rel(0);
+    final Window windowRel = call.rel(1);
+
+    // Get the window all groups
+    List<Window.Group> windowGroups = windowRel.groups;
+
+    // The window may have multi groups,now we could only
+    // deal one group case,so that we could know the partition keys.
+    if (windowGroups.size() != 1) {
+      return;
+    }
```

Comment:
Is this a temporary (implementation) limitation or a permanent (semantic) limitation? IF there are multiple windows can't we consider the intersection of partition keys?

Reply:
My current thinking is that it is a permanent limitation, because it is somewhat difficult to find the common intersection of multiple groups.And like other engine,such as Presto/Trino/StarRocks, they handle window with a single group.

My idea is that in the future I would like to add a rule to split a Window with multiple Groups into multiple Window with a single Group,so the FilterWindowTransposeRule could also push the condition to below the bottom Window.

But this is not the final conclusion. Once I have an idea in the future, it will be more versatile to handle multiple Group partitions at the same time, and if users using Calcite can use this optimization, I will continue to improve this optimization rule.

WDYT?